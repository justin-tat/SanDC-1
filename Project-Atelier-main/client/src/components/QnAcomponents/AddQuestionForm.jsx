import React from 'react';
import axios from 'axios';

class AddQuestionForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isValid: false,
      questionBody: '',
      nickname: '',
      email: '',
      isWarningShown: false

    };
    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleValidation = this.handleValidation.bind(this);
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.value;
    const name = target.name;
    this.setState({
      [name]: value
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    var productId = this.props.productId;
    var body = this.state.questionBody;
    var nickname = this.state.nickname;
    var email = this.state.email;
    var validationResult = this.handleValidation(body, nickname, email);

    if (validationResult) {
      this.setState({
        isValid: true,
        isWarningShown: false
      });
      this.props.addQuestion(productId, body, nickname, email);
      this.props.closeForm();

    } else {
      this.setState({
        isWarningShown: true
      });
    }
  }
  handleValidation(question, nick, email) {
    if (question.length === 0) {
      return false;
    } else if (nick.length === 0) {
      return false;
    } else if (!(/^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/.test(email))) {
      return false;
    } else {
      return true;
    }
  }


  render() {
    let warning;
    if (this.state.isWarningShown) {
      warning = <div className='qna-add-question-form-warning-message'>You didn't properly filled all the fields</div>;
    } else {
      warning = <div></div>;
    }
    return (
      <div className='qna-add-new-question-form'>
        <div className='qna-add-question-form-wrapper'>
          {/* <div className = 'qna-add-question-main-title'>Ask a question</div> */}
          <div className ='qna-add-question-subtitle'>Ask a question about the {this.props.name}</div>
          <form onSubmit={this.handleSubmit}>
            <label>
              <div className ='qna-add-question-form-your-question'>Your question*</div>
              <input
                className='qna-add-question-input-questionbody'
                name='questionBody'
                id='questionBody'
                type='text'
                maxLength='1000'
                value={this.state.questionBody}
                onChange={this.handleInputChange}
              />
            </label>
            <br />
            <label>
              <div className='qna-add-question-form-your-nickname'>What's your nickname?*</div>
              <input
                className='qna-add-question-input-nickname'
                name='nickname'
                type='text'
                maxLength='60'
                placeholder='Example: jackson11'
                value={this.state.nickname}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-question-form-warning-1'>
               For privacy reasons, do not use your full name or email address
              </div>
            </label>
            <br />
            <label>
              <div className='qna-add-question-form-your-email'>
              Your email?*
              </div>
              <input
                className='qna-add-question-form-input-email'
                name='email'
                type='text'
                maxLength='60'
                placeholder='Example: sample@email.com'
                value={this.state.email}
                onChange={this.handleInputChange}
              />
              <br />
              <div className='qna-add-question-form-warning-2'>
               For authentication reasons, you will not be emailed
              </div>
            </label>
            <br />
            {warning}
            <input type="submit" className="qna-add-question-form-submit" value="Submit" onClick = {(e)=>this.handleSubmit(e)} />
            <button className='qna-add-question-form-close' onClick={()=> this.props.closeForm()}>Close without submitting</button>
          </form>
        </div>
      </div>

    );
  }
}

export default AddQuestionForm;