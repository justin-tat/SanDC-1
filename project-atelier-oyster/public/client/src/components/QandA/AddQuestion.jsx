import React from 'react';
import axios from 'axios';

class AddQuestion extends React.Component {
  constructor(props) {
  super(props)
  this.state = {
    show: false
  };
  this.showModal = this.showModal.bind(this);
  this.hideModal = this.hideModal.bind(this);
  this.addQuestion = this.addQuestion.bind(this);
  }

  showModal (e) {
    this.props.clickTracker('Add Question Button', 'AddQuestion.jsx');
    this.setState({ show: true })
  }

  hideModal (e) {
    this.props.clickTracker('Submit Question Button', 'AddQuestion.jsx');
    this.setState({ show: false })
  }

  addQuestion (e) {
    e.preventDefault();
    let body = e.target[0].value;
    let name = e.target[1].value;
    let email = e.target[2].value;
    let id = this.props.id;

    axios.post(`/addQuestion`, {data: {body: body, name: name, email: email, product_id: id}})
    .then((result) => {
      console.log('Question was added');
      this.props.update(id, 1, 100);
    })
    .catch((error) => {
      console.error(error);
    })
    this.hideModal();
  }

  render () {
    if (this.state.show === false) {
      return(
        <span className='addQuestion stuff' onClick={this.showModal}>Add A Question +</span>
      )
    } else {
      return(
        <div className='modalBackground'>
          <div className='modalContainer'>
            <h3>Add Your Question</h3>
            <form className='newQForm' onSubmit={this.addQuestion}>
              <label className='bodylabel' for='body'>Your Question*</label>
              <input className='bodyInput' type='text' placeholder='Your Question Here' ref='body' name='body' required></input><br></br>
              <label className='nameLabel' for='name'>What is your nickname*</label>
              <input className='nameInput' type='text' placeholder='Example: jackson11!' ref='name' name='name' required></input><br></br>
              <div className='nameWarning'>For privacy reasons, do not use your full name or email address</div>
              <label className='emailLabel' for='email'>Your email*</label>
              <input className='emailInput' type='email' placeholder='Why did you like the product of not?' ref='email' name='email' required></input><br></br>
              <div className='emailWarning'>For authentication reasons, you will not be emailed<br></br>Fields marked with * are erquired</div>
              <input type='submit' value='submit'></input>
            </form>
          </div>
        </div>
      )
    }
  };
}

export default AddQuestion;