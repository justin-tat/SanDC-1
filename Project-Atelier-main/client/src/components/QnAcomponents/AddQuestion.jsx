import React from 'react';
import AddQuestionForm from './AddQuestionForm.jsx';

class AddQuestion extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isModalShown: false,
      isAddButtonShown: true,
      productId: this.props.productId
    };
    this.clickOnAddQuestion = this.clickOnAddQuestion.bind(this);
    this.closeForm = this.closeForm.bind(this);

  }

  clickOnAddQuestion() {
    this.setState({
      isModalShown: true,
      isAddButtonShown: false
    });
    this.props.checkForm();
  }

  closeForm() {
    console.log('close form is triggered');
    this.setState({
      isModalShown: false,
      isAddButtonShown: true
    });
  }


  render() {
    let qnaAskQuestionModal,
      questionButton;
    if (this.state.isModalShown) {
      qnaAskQuestionModal = 'qna-ask-question-modal-shown';
    } else {
      qnaAskQuestionModal = 'qna-ask-question-modal-hidden';
    }
    // let modal,

    // if (this.state.isModalShown) {
    //   modal = <div></div>;
    // } else {
    //   modal = <div></div>;
    // }
    if (this.state.isAddButtonShown) {
      questionButton = <button className='qna-add-question-button' onClick={()=>this.clickOnAddQuestion()}>ADD A QUESTION</button>;
    } else {
      questionButton = <div></div>;
    }
    return (
      <div className='add-question-parent'>
        {questionButton}
        <div className={qnaAskQuestionModal}><AddQuestionForm
          name={this.props.name}
          productId={this.props.productId}
          addQuestion={this.props.addQuestion}
          closeForm={this.closeForm}
        />
        </div>
      </div>
    );
  }
}

export default AddQuestion;