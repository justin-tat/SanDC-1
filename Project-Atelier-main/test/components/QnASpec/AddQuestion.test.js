import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

import React from 'react';
import { shallow, mount, render, ShallowWrapper } from 'enzyme';
//example data
import exampleQuestions from '../../../example/questions.js';

import AddQuestionForm from '../../../client/src/components/QnAcomponents/AddQuestionForm.jsx';
import AddQuestion from '../../../client/src/components/QnAcomponents/AddQuestion.jsx';



describe('Rendering add question parent component', function() {
  xit('should render without throwing an error', function() {
    expect(shallow(<AddQuestion />).contains( <div className='qna-add-new-question-form'></div>
    )).toBe(true);
  });

  it('should be selectable by class', function() {
    expect(shallow(<AddQuestion />).is('.add-question-parent')).toBe(true);
  });

  it('should mount in a full DOM', function() {
    expect(mount(<AddQuestion />).find('.add-question-parent').length).toBe(1);
  });

  xit('should render to static HTML', function() {
    var text = 'Ask a question about the Your question*What\'s your nickname?*For privacy reasons, do not use your full name or email addressYour email?*For authentication reasons, you will not be emailed';
    expect(render(<AddQuestion />).text()).toEqual(text);
  });

  test('invokes clickOnAddQuestion() and correctly changing state if clicking on button \'Add question\' ', () => {
    const mockCheckForm = jest.fn();
    const wrapper = shallow(<AddQuestion checkForm={mockCheckForm} />);
    const spy = jest.spyOn(wrapper.instance(), 'clickOnAddQuestion');
    wrapper.find('.qna-add-question-button').simulate('click');
    expect(spy).toHaveBeenCalled();
    //console.log(wrapper.state());
    expect(wrapper.state().isModalShown).toEqual(true);
    expect(wrapper.state().isAddButtonShown).toEqual(false);

  });



  xtest('child component correctly changes state of  parent component ', () => {
    const parent = shallow(<AddQuestion />);
    const spy = jest.spyOn(parent.instance(), 'closeForm');

    //imitating adding new question
    parent.find('.qna-add-question-button').simulate('click');
    expect(parent.state().isModalShown).toEqual(true);
    expect(parent.state().isAddButtonShown).toEqual(false);

    const formEventMocked = { preventDefault: jest.fn() };
    const passedMock1 = jest.fn();
    const wrapper = shallow(<AddQuestionForm addQuestion={passedMock1} closeForm={spy} />);

    const state = {
      questionBody: 'This is a question from tests',
      nickname: 'gandalf',
      email: 'example@email.com'
    };
    wrapper.setState(state);

    wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    //
    expect (spy).toHaveBeenCalled();
    expect(parent.state().isModalShown).toEqual(false);
    expect (parent.state().isAddButtonShown).toEqual(true);

  });

});

describe('Add new question form', function() {

  test('click on submit button without filling the form', () => {
    const formEventMocked = { preventDefault: jest.fn() };
    const alertSpy = window.alert;
    window.alert = () => {}; // provide an empty implementation for window.alert

    const wrapper = shallow(<AddQuestionForm/>);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');
    //wrapper.find('[type="submit"]').simulate('submit');
    expect(wrapper.find('form')).toHaveLength(1);
    //wrapper.find('form').simulate('submit', formEventMocked);
    wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    expect(formEventMocked.preventDefault).toBeCalledTimes(1);
    expect(spy3).toBeCalledTimes(1);
    expect(spy1).not.toHaveBeenCalled();

  });

  test('click on submit button when form not filled properly', () => {
    const formEventMocked = { preventDefault: jest.fn() };
    const alertSpy = window.alert;
    window.alert = () => {}; // provide an empty implementation for window.alert

    const wrapper = shallow(<AddQuestionForm/>);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');

    const state = {
      questionBody: 'This is a question from tests',
      nickname: 'gandalf',
      email: ''
    };
    wrapper.setState(state);
    expect(wrapper).toMatchSnapshot();
    //expect(wrapper.find('form')).toHaveLength(1);

    wrapper.find('[type="submit"]').simulate('click', formEventMocked);

    expect(formEventMocked.preventDefault).toBeCalledTimes(1);
    expect(spy3).toBeCalledTimes(1);

  });

  test('allows to type in the form fields', () => {
    const formEventMocked = { preventDefault: jest.fn() };
    const alertSpy = window.alert;
    window.alert = () => {}; // provide an empty implementation for window.alert

    const wrapper = mount(<AddQuestionForm/>);
    //wrapper.find('input').simulate('change', {target: {value: 'Your new Value'}});
    const question = wrapper.find({name: 'questionBody'});
    question.props().value = 'foo';
    //question.simulate('change', { target: { value: 'bla' } });
    //console.log(question.debug());
    //wrapper.update();
    //wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    expect(question.props().value).toBeTruthy();

  });

  test('handle validation should return false without input', () => {
    const component = new AddQuestionForm;
    expect(component.handleValidation('', '', '')).toEqual(false);
  });

  test('handle validation should return false without question', () => {
    const component = new AddQuestionForm;
    expect(component.handleValidation('', 'aragorn', 'kings@dunedain.com')).toEqual(false);
  });

  test('handle validation should return false without nickname', () => {
    const component = new AddQuestionForm;
    expect(component.handleValidation('for frodo', '', 'kings@dunedain.com')).toEqual(false);
  });

  test('handle validation should return false without email', () => {
    const component = new AddQuestionForm;
    expect(component.handleValidation('for frodo', 'bill the pony', '')).toEqual(false);
  });

  test('handle validation should return false with email in wrong format', () => {
    const component = new AddQuestionForm;
    expect(component.handleValidation('for frodo', 'bill the pony', 'dummyemail')).toEqual(false);
  });

  test('handle validation should return true with proper arguments', () => {
    const component = new AddQuestionForm;
    expect(component.handleValidation('for frodo', 'bill the pony', 'ponies@example.com')).toEqual(true);
  });

  test('isValid if true should invoked passed functions', () => {

    const formEventMocked = { preventDefault: jest.fn() };

    const passedMock1 = jest.fn();
    const passedMock2 = jest.fn();
    const wrapper = shallow(<AddQuestionForm addQuestion={passedMock1} closeForm={passedMock2} />);
    const spy1 = jest.spyOn(wrapper.instance(), 'handleInputChange');
    const spy2 = jest.spyOn(wrapper.instance(), 'handleSubmit');
    const spy3 = jest.spyOn(wrapper.instance(), 'handleValidation');

    const state = {
      questionBody: 'This is a question from tests',
      nickname: 'gandalf',
      email: 'example@email.com'
    };
    wrapper.setState(state);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('[type="submit"]').simulate('click', formEventMocked);
    expect(wrapper.state().isValid).toEqual(true);
    expect(spy3).toBeCalledTimes(1);
    expect(passedMock1).toBeCalledTimes(1);
    expect(passedMock2).toBeCalledTimes(1);

  });

  it('should render without throwing an error', function() {
    expect(shallow(<AddQuestionForm />).contains(<div className ='qna-add-question-form-your-question'>Your question*</div>)).toBe(true);
  });

  it('should be selectable by class', function() {
    expect(shallow(<AddQuestionForm />).is('.qna-add-new-question-form')).toBe(true);
  });

  it('should mount in a full DOM', function() {
    expect(mount(<AddQuestionForm />).find('.qna-add-new-question-form').length).toBe(1);
  });

  it('should render to static HTML', function() {
    var text = 'Ask a question about the Your question*What\'s your nickname?*For privacy reasons, do not use your full name or email addressYour email?*For authentication reasons, you will not be emailedClose without submitting';
    expect(render(<AddQuestionForm />).text()).toEqual(text);
  });

  it('should have properties at state', () => {
    const wrapper = mount(<AddQuestionForm
    />);
    expect(wrapper.state().isValid).toEqual(false);
    expect(wrapper.state().questionBody).toEqual('');
    expect(wrapper.state().nickname).toEqual('');
    expect(wrapper.state().email).toEqual('');
  });

  it('allows to close form without submitting', () => {
    const mock = jest.fn();
    const wrapper = mount(<AddQuestionForm closeForm={mock}
    />);
    wrapper.find('.qna-add-question-form-close').simulate('click');
    expect(mock).toHaveBeenCalled();

  });

});