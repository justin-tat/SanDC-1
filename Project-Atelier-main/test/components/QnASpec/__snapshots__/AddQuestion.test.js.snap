// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add new question form click on submit button when form not filled properly 1`] = `ShallowWrapper {}`;

exports[`Add new question form isValid if true should invoked passed functions 1`] = `ShallowWrapper {}`;
